### PR TITLE
Added `ignore` config option

### PR DESCRIPTION
**Pull request by k2snowman69**
Goal is to allow limiting of what files sortier runs on

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -41,6 +41,21 @@ function findAndRunSortier(
...
 ) {
+  const fileName = document.fileName;
+
+  let included = true;
+  const inclusions = vscode.workspace
+    .getConfiguration(extensionName)
+    .get<Array<string>>("includes");
+  if (inclusions != null && inclusions.length > 0) {
+    let mic = micromatch;
+    included = mic.isMatch(fileName, inclusions);
```
Comment (k2snowman69) [inline src/extension.ts]: For some reason this fails when running with a variable but when running with a direct string everything is fine in windows....

```
mic.isMatch(fileName, inclusions)
false
fileName
"d:\Projects\react-ui-base\dialog\.lintstagedrc.js"
inclusions
Array(1) ["*.js"]
mic.isMatch("d:\Projects\react-ui-base\dialog\.lintstagedrc.js", ["*.js"])
true
```

Comment (k2snowman69) [reply src/extension.ts]: Possible solution maybe using the `ignore` package instead of minimatch

**Comment by k2snowman69** [comment]
@TikiTDO I couldn't get `minimatch`, `micromatch` and several others to work but I was able to get `ignore` to work which works just like a gitignore file. Mind giving this pull request a try and if it works out for you signing off and closing out your other pull request?

**Comment by TikiTDO** [comment]
@k2snowman69 How do you normally test these? I'm not super familiar with vscode extension development, so I previously just went in and modified the extension code in place. Is there a simple way to just point it at a branch and say "use this code"?

File: package.json [hunk 2]
```diff
@@ -19,6 +19,15 @@
     "configuration": [
       {
         "properties": {
+          "sortier.ignore": {
+            "default": [],
+            "description": "Array of glob patterns to ignore",
+            "items": {
+              "title": "Glob to include",
```
Comment (TikiTDO) [inline package.json]:
```suggestion
              "title": "Glob to ignore",
```

File: src/extension.ts [hunk 3]
```diff
@@ -41,6 +41,21 @@ function findAndRunSortier(
   document: vscode.TextDocument,
   messageIfFileNotSupported: boolean = true
 ) {
+  const fileName = document.fileName;
+
+  let ignored = true;
```
Comment (TikiTDO) [inline src/extension.ts]: Looks like this will ignore everything if you don't specify an ignore param.

File: src/extension.ts [hunk 4]
```diff
@@ -41,6 +41,21 @@ function findAndRunSortier(
...
+  const inclusions = vscode.workspace
```
Comment (TikiTDO) [inline src/extension.ts]: The `inclusions` name is a bit confusing here, something like `patternsToIgnore` would make it easier when returning to this code in the future.

File: src/extension.ts [hunk 5]
```diff
@@ -41,6 +41,21 @@ function findAndRunSortier(
...
+    .getConfiguration(extensionName)
+    .get<Array<string>>("ignore");
+  if (inclusions != null && inclusions.length > 0) {
+    const ig = ignore().add(inclusions);
+    ignored = ig.ignores(fileName);
```
Comment (TikiTDO) [inline src/extension.ts]: You could do the return inline here, this way you wouldn't need the `ignored` variable at all.

**Comment by k2snowman69** [comment]
The way I usually test this is:

1. Pull down the repo (`git clone https://github.com/snowcoders/sortier-vscode`)
2. Checkout the branch (`git checkout k2snowman69/includesProperty`)
3. Open the project in VS Code
4. Run the debugger in VS Code and it will open a workspace with the debug version of the plugin installed

Thanks for the code comments as well, I'll get those addressed :-)

**Comment by TikiTDO** [comment]
Maybe approaching it from the `include` direction would be better? 

Right now my config looks like this: `"sortier.ignore": ["**/*", "!**/*.js",  "!**/*.jsx",  "!**/*.ts",  "!**/*.tsx"]`. I can definitely see people getting confused and posting issues about trying to follow the ignore logic in a config file.

File: src/extension.ts (hunk 1 above)
Comment (k2snowman69) [reply src/extension.ts]: Moved to using ignore

**Comment by k2snowman69** [comment]
I guess a better question here is actually what are you trying to fix with this includes scenario? Because with that ignore statement you're pretty much including everything that sortier runs on other than html and css files. 

Is it the message "Sortier threw an error: Could not find language support for file - <filepath>"? If so then that's a much simpler fix on the "onSave"  handler and making sure the file selected is supported.

**Comment by TikiTDO** [comment]
That was just what I was using for testing. For the actual use case there's a bunch of files that I'd add to a workspace config... Which now that I think about it fits the use case of `gitignore` a lot more.

**Comment by k2snowman69** [comment]
Would it be better to have a `.sortierignore` file instead of having this in the vs code config?

**Comment by TikiTDO** [comment]
Yeah, that would probably be even better actually. We have a CI check to ensure sortier was run on the codebase, and having a generic config file would make that even cleaner.

**Comment by k2snowman69** [comment]
Cool, in that case that pull request would be more generic in the https://github.com/snowcoders/sortier repo but I think it makes sense as it would apply more generally to everyone. It would also allow for less complicated git hooks which would be nice.

I'll leave this open as a reference since the code would be fairly similar other than instead of pulling the values from vscode, we would be pulling it from `.sortierignore`.

**Comment by k2snowman69** [comment]
Looks like we're unlucky and when we run sortier via the plugin, it runs in the context of vscode not the context of the workspace... we're going to have to:
a. Duplicate the ignore logic that was added to the sortier package
b. Add a config option in the CLI that allows us to specify the sortierignore file location and pass in the full location of the sorterignore file found in the workspace to the CLI via the `format` function's options in the plugin.

Prettier also allows you to provide a uri as a option to say where the ignore file is... I'm curious what other programs do...

**Comment by TikiTDO** [comment]
The way I've seen other tools handle it is to simply walk up the tree from the file being processed until they find a relevant config file.

**Comment by k2snowman69** [comment]
So walk up from the current file being sortier'ed? I do like that solution...

**Comment by k2snowman69** [comment]
Implemented walking up from the file being sortier'ed
https://github.com/snowcoders/sortier/pull/414/files

**Comment by k2snowman69** [comment]
Closing in favor of https://github.com/snowcoders/sortier-vscode/pull/126

**Pull request closed without merging**